src/api/source/mvt.js

```diff
@@ -0,0 +1,175 @@
+import * as rsys from '../../client/rsys';
+import Dataframe from '../../core/dataframe';
+import * as Protobuf from 'pbf';
+import { VectorTile } from '@mapbox/vector-tile';
+import { decodeLines, decodePolygons } from '../../client/mvt/feature-decoder';
+import TileClient from './TileClient';
+import Base from './base';
+
+const geometryTypes = {
+    UNKNOWN: 'unknown',
+    POINT: 'point',
+    LINE: 'line',
+    POLYGON: 'polygon'
+};
+
+export default class MVT extends Base {
+
+    /**
+     * Create a carto.source.MVT.
+     *
+     * @param {object} data - A MVT data object
+     * @param {object} metadata - A carto.source.mvt.Metadata object
+     *
+     * @example
+     * const metadata = new carto.source.mvt.Metadata([{ type: 'number', name: 'total_pop'}])
+     * new carto.source.MVT("https://{server}/{z}/{x}/{y}.mvt", metadata);
+     *
+     * @fires CartoError
+     *
+     * @constructor MVT
+     * @extends carto.source.Base
+     * @memberof carto.source
+     * @api
+     */
+    // FIXME, multi subdomain
+    constructor(templateURL, metadata = { columns: [] }) {
+        super();
+        this._templateURL = templateURL;
+        this._metadata = metadata;
+        this._tileClient = new TileClient(templateURL);
+        this._categoryStringToIDMap = {};
+        this._numCategories = 0;
+    }
+
+    _clone() {
+        return new MVT(this._templateURL, JSON.parse(JSON.stringify(this._metadata)));
+    }
+
+    bindLayer(addDataframe, dataLoadedCallback) {
+        this._tileClient.bindLayer(addDataframe, dataLoadedCallback);
+    }
+
+    async requestMetadata() {
+        return this._metadata;
+    }
+
+    requestData(viewport) {
+        return this._tileClient.requestData(viewport, this.responseToDataframeTransformer.bind(this));
+    }
+
+    async responseToDataframeTransformer(response, x, y, z) {
+        const mvt_extent = 4096;
+        const arrayBuffer = await response.arrayBuffer();
+        if (arrayBuffer.byteLength == 0 || response == 'null') {
+            return { empty: true };
+        }
+        const tile = new VectorTile(new Protobuf(arrayBuffer));
+        const mvtLayer = tile.layers[Object.keys(tile.layers)[0]];
+
+        const { points, featureGeometries, properties } = this._decodeMVTLayer(mvtLayer, this._metadata, mvt_extent);
+
+
+        const rs = rsys.getRsysFromTile(x, y, z);
+        const dataframeGeometry = this._metadata.geomType == geometryTypes.POINT ? points : featureGeometries;
+        const dataframe = this._generateDataFrame(rs, dataframeGeometry, properties, mvtLayer.length, this._metadata.geomType);
+        return dataframe;
+    }
+
+
+    _decodeMVTLayer(mvtLayer, metadata, mvt_extent) {
+        let points;
+        if (metadata.geomType == geometryTypes.POINT) {
```

Comment:
I think this early check is not needed, because it is also done again in the loop.

Reply:
:+1:

Reply:
I thought it was redundant, but, after changing it, it seems it's not. The reason is that the second `if` is executed only when the geometry type wasn't defined previously. This makes this first check necessary, of course, additional logic can be used to remove this `if` and make the second `if` work in both cases, but... I think it's basically the same.

Reply:
ok!